File: src/main/java/net/fabricmc/mante/discordian/AccountLinkManager.java
package net.fabricmc.mante.discordian;

import net.dv8tion.jda.api.entities.Member;
import net.minecraft.registry.Registries;
import net.minecraft.registry.RegistryKey;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.text.Text;
import net.minecraft.text.TextColor;
import net.minecraft.util.Formatting;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.GameMode;
import net.minecraft.world.World;

import java.util.ArrayList;
import java.util.UUID;

public class AccountLinkManager {
    public ArrayList<AccountLinkDetails> pending;

    public boolean codeAlreadyExists(String code) {
        for (AccountLinkDetails details : pending) {
            if (details.code.equals(code)) {
                return true;
            }
        }

        return false;
    }

    private String getFourDigitSequence() {
        return String.valueOf(Discordian.RANDOM.nextInt(1000, 10000));
    }

    public String generateCode() {
        String code;

        do {
            code = getFourDigitSequence();
        }
        while (codeAlreadyExists(code));

        return code;
    }

    public AccountLinkManager() {
        pending = new ArrayList<>();
    }

    public void addPending(String uuid, String code) {
        pending.add(new AccountLinkDetails(uuid, code));
    }

    public void removePending(String code) {
        pending.removeIf(accountLinkDetails -> accountLinkDetails.code.equals(code));
    }

    public void link(String id, String code) {
        for (AccountLinkDetails account : pending) {
            if (code.equals((account.code))) {
                Discordian.configManager.addLinkedAccount(account.uuid, id);
                removePending(code);

                ServerPlayerEntity player = Discordian.server.getPlayerManager().getPlayer(UUID.fromString(account.uuid));

                if (player != null) {
                    Member member = Discordian.getMember(id);
                    if (member == null)
                        return;

                    player.sendMessage(
                            Text.literal("Your account has been linked to ").styled(style -> style.withColor(TextColor.fromFormatting(Formatting.GREEN)))
                                    .append(Text.literal(member.getUser().getAsTag()).styled(style -> style.withColor(TextColor.fromFormatting(Formatting.YELLOW)))),
                            false);

                    if (Discordian.configManager.config.get("require_link_to_play").getAsBoolean()) {
                        player.changeGameMode(GameMode.DEFAULT);
                        ServerWorld spawnWorld = Discordian.server.getWorld(player.getSpawnPointDimension());
                        BlockPos spawnPos = player.getSpawnPointPosition();
                        if (spawnPos == null) {
                            spawnPos = Discordian.server.getOverworld().getSpawnPos();
                        }
                        player.teleport(spawnWorld, spawnPos.getX(), spawnPos.getY(), spawnPos.getZ(), 0.0f, 0.0f);
                    }
                }

                return;
            }
        }
    }

    public boolean isLinked(String uuid) {
        return Discordian.configManager.config.get("linked_accounts").getAsJsonObject().has(uuid);
    }

    public String uuidToId(String uuid) {
        return Discordian.configManager.config.get("linked_accounts").getAsJsonObject().has(uuid) ? Discordian.configManager.config.get("linked_accounts").getAsJsonObject().get(uuid).getAsString() : "";
    }
}

File: src/main/java/net/fabricmc/mante/discordian/AccountLinkManager.kt
package net.fabricmc.mante.discordian

import kotlinx.coroutines.launch
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.runBlocking
import net.minecraft.text.Text
import net.minecraft.text.TextColor
import net.minecraft.util.Formatting
import net.minecraft.world.GameMode
import java.util.*

class AccountLinkManager {
	var pending: ArrayList<AccountLinkDetails> = ArrayList()

	fun codeAlreadyExists(code: String): Boolean {
		for (details in pending) {
			if (details.code == code) {
				return true
			}
		}
		return false
	}

	private val fourDigitSequence: String
		get() = java.lang.String.valueOf(Discordian.random.nextInt(1000, 10000))

	fun generateCode(): String {
		var code: String
		do {
			code = fourDigitSequence
		} while (codeAlreadyExists(code))
		return code
	}

	fun addPending(uuid: String, code: String) {
		pending.add(AccountLinkDetails(uuid, code))
	}

	fun removePending(code: String) {
		pending.removeIf { accountLinkDetails: AccountLinkDetails -> accountLinkDetails.code == code }
	}

	fun link(id: String, code: String) {
		val account = pending.find { it.code == code } ?: return

		Config.inst().addLinkedAccount(account.uuid, id)
		removePending(code)

		Discordian.kordScope.launch {
			val player =
				Discordian.server.playerManager.getPlayer(UUID.fromString(account.uuid)) ?: return@launch
			val member =
				Discordian.channel?.getGuild()?.let { Discordian.kord.getMember(id, it) } ?: return@launch

			player.sendMessage(
				Text.literal("Your account has been linked to ").styled {
					it.withColor(TextColor.fromFormatting(Formatting.GREEN))
				}
					.append(Text.literal(member.username).styled {
						it.withColor(TextColor.fromFormatting(Formatting.YELLOW))
					}),
				false
			)
			if (Config.inst().requireLinkToPlay) {
				player.changeGameMode(GameMode.DEFAULT)
				val spawnWorld =
					Discordian.server.getWorld(player.spawnPointDimension) ?: Discordian.server.overworld
				val spawnPos = player.spawnPointPosition ?: Discordian.server.overworld.spawnPos

				player.teleport(
					spawnWorld,
					spawnPos.x.toDouble(),
					spawnPos.y.toDouble(),
					spawnPos.z.toDouble(),
					0.0f,
					0.0f
				)
			}
		}
	}
}